### PR TITLE
Add Decimal256Builder and Decimal256Array

### LINKED ISSUE #1999: Implement Decimal256Builder and Decimal256Array
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

In order to support for Decimal with 256 bits of precision, we need to implement Decimal256Builder and Decimal256Array.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1999.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by viirya** [comment]
Marked as draft now and I will add some tests and dedup some codes.

**Comment by liukun4515** [comment]
If it is ready, please at me. @viirya

File: arrow/src/array/builder/decimal_builder.rs [hunk 1]
```diff
@@ -197,4 +258,42 @@ mod tests {
...
+        builder.append_null().unwrap();
+
+        bytes = vec![255; 32];
+        let value = Decimal256::try_new_from_bytes(40, 6, bytes.as_slice()).unwrap();
+        builder.append_value(&value).unwrap();
+
+        bytes = vec![0; 32];
+        bytes[0..16].clone_from_slice(&0_i128.to_le_bytes());
+        bytes[15] = 128;
+        let value = Decimal256::try_new_from_bytes(40, 6, bytes.as_slice()).unwrap();
```
Comment (alamb) [inline arrow/src/array/builder/decimal_builder.rs:280]: I think we should add a test for appending a `value` that has a precision/scale other than `40. 6`, as I mentioned above

File: arrow/src/array/builder/decimal_builder.rs [hunk 2]
```diff
@@ -154,10 +166,59 @@ impl ArrayBuilder for DecimalBuilder {
...
+
+    /// Appends a byte slice into the builder.
+    ///
+    /// Automatically calls the `append` method to delimit the slice appended in as a
+    /// distinct array element.
+    #[inline]
+    pub fn append_value(&mut self, value: &Decimal256) -> Result<()> {
+        let value_as_bytes = value.raw_value();
+
+        if self.builder.value_length() != value_as_bytes.len() as i32 {
```
[the remark this answers is not in the document]
Comment (viirya) [reply arrow/src/array/builder/decimal_builder.rs:190]: Good point. I think we should. It is a cheap check too so I think okay.

File: arrow/src/array/builder/decimal_builder.rs [hunk 3]
```diff
@@ -42,8 +44,18 @@ pub struct DecimalBuilder {
     value_validation: bool,
 }
 
+/// Array Builder for [`Decimal256Array`]
+///
+/// See [`Decimal256Array`] for example.
+#[derive(Debug)]
+pub struct Decimal256Builder {
+    builder: FixedSizeListBuilder<UInt8Builder>,
```
[the remark this answers is not in the document]
Comment (HaoYang670) [reply arrow/src/array/builder/decimal_builder.rs:52]: Create https://github.com/apache/arrow-rs/issues/2026 to track this.

**Comment by alamb** [comment]
@tustvold  is this one good to go (for arrow 18?)

**Comment by alamb** [comment]
I see now that @liukun4515  asked to be pinged on this PR for review, which @viirya did but then I merged it prior to @liukun4515 having a chance to review. 

@liukun4515  I will make the 18.0.0 release candidate but if you find issues with the API or some other reason we should not release 18.0.0 with this change in it, please let me know and we can make changes prior to release.

File: arrow/src/array/builder/decimal_builder.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (liukun4515) [reply arrow/src/array/builder/decimal_builder.rs:52]: I also confused about that we use the the `FixedSizeListBuilder` to store decimal data in the decimal builder.

**Comment by liukun4515** [comment]
> I see now that @liukun4515 asked to be pinged on this PR for review, which @viirya did but then I merged it prior to @liukun4515 having a chance to review.
> 
> @liukun4515 I will make the 18.0.0 release candidate but if you find issues with the API or some other reason we should not release 18.0.0 with this change in it, please let me know and we can make changes prior to release.



> I see now that @liukun4515 asked to be pinged on this PR for review, which @viirya did but then I merged it prior to @liukun4515 having a chance to review.
> 
> @liukun4515 I will make the 18.0.0 release candidate but if you find issues with the API or some other reason we should not release 18.0.0 with this change in it, please let me know and we can make changes prior to release.

Sorry for the later reply, I am busy with other task.
This pr looks great for me.
And from https://github.com/apache/arrow-rs/pull/2000#discussion_r916386516, it's better to refactor the builder and generate the array data from `fixed list builder` to `fixed binary builder`.

**Pull request merged**